### ISSUE TITLE
Shuffle batches across workers

### ISSUE THREAD

**Issue by platers** [pytorch/data#1140]
### 🚀 The feature

I have a Dataloader with n workers. My understanding is that each worker constructs a full batch independently, which is then served by the dataloader. My samples are large, so I cannot increase the shuffle buffer size in each worker. Is there a way to perform the batching and shuffling only in the main process? 

### Motivation, pitch

This would improve shuffling for a fixed memory usage.

### Alternatives

I tried having an inner dataloader with n workers that produces samples, and an outer dataloader that shuffles and batches them. I couldnt get the inner dataloader to use n workers.

**Comment by ejguan** [comment]
You use a non-replicable DataPipe to indicate where your want to put the DataPipe in the main process. For your use case, you can do following:
```py
class NonReplicableDataPipe(IterDataPipe):
    def __init__(self, dp):
        self.dp = dp
    def __iter__(self):
        yield from self.dp
    def is_replicable(self):
        return False

if __name__ == "__main__":
    dp = ...
    dp = NonReplicable(dp)
    dp = dp.shuffle().batch()
    rs = MultiProcessingReadingService(xxx)
    dl = DataLoader2(dp, reading_service=rs)

```


For documentation issue, it's been tracked in https://github.com/pytorch/data/issues/1079

**Comment by platers** [comment]
Thanks for the help!
